nni/algorithms/compression/v2/pytorch/utils/evaluator.py

```diff
@@ -766,3 +774,175 @@ def evaluate(self) -> float | None | Tuple[float, Dict[str, Any]] | Tuple[None,
+        self._ori_trainer_attr['optimizer.step'] = self.trainer.optimizer.step
+
+    def unbind_model(self):
+        if self.model:
+            self.revert_loss()
+            self.revert_optimizer_step()
+            self.remove_all_hooks()
+            self._ori_trainer_attr.pop('optimizer.step', None)
+            self.trainer.optimizer = None
+            self._param_names_map = None
+            self._ori_trainer_attr.pop('compute_loss', None)
+            self.trainer = None
+            self.model = None
+        else:
+            _logger.warning('Did not bind any model, no need to unbind model.')
+
+    def patch_loss(self, patch: Callable[[Tensor], Tensor]):
+        old_compute_loss = self.trainer.compute_loss
+
+        def patched_compute_loss(_, model: Any, inputs: Any, return_outputs: bool = False):
+            result = old_compute_loss(model, inputs, return_outputs)
+            if return_outputs:
+                return patch(result[0]), result[1]
+            else:
+                return patch(result)
+
+        self.trainer.compute_loss = types.MethodType(patched_compute_loss, self.trainer)
+
+    def revert_loss(self):
+        self.trainer.compute_loss = self._ori_trainer_attr['compute_loss']
+
+    def patch_optimizer_step(self, before_step_tasks: List[Callable], after_step_tasks: List[Callable]):
+        assert self.trainer.optimizer is not None
+        old_step = self.trainer.optimizer.step
+
+        def patched_step(_, *args, **kwargs):
+            for task in before_step_tasks:
+                task()
+            # call origin optimizer step method
+            output = old_step(*args, **kwargs)
+            for task in after_step_tasks:
+                task()
+            return output
+
+        self.trainer.optimizer.step = types.MethodType(patched_step, self.trainer.optimizer)
+
+    def revert_optimizer_step(self):
+        assert self.trainer.optimizer is not None
+        self.trainer.optimizer.step = self._ori_trainer_attr['optimizer.step']
+
+    def train(self, max_steps: int | None = None, max_epochs: int | None = None):
+        assert self.model is not None
+        ori_steps, ori_epochs = self.trainer.args.max_steps, self.trainer.args.num_train_epochs
+        if max_epochs is not None:
+            self.trainer.args.num_train_epochs = max_epochs
+        if max_steps is not None:
+            self.trainer.args.max_steps = max_steps
+        self.trainer.lr_scheduler = self._lr_scheduler_helper.call(self.trainer.optimizer) if self._lr_scheduler_helper else None
+
+        self.trainer.train()
+
+        self.trainer.lr_scheduler = None
+        self.trainer.args.max_steps, self.trainer.args.num_train_epochs = ori_steps, ori_epochs
+
+    def finetune(self):
+        self.train()
+
+    def evaluate(self) -> float | None | Tuple[float, Dict[str, Any]] | Tuple[None, Dict[str, Any]]:
+        return self.trainer.evaluate()
+
+    def get_dummy_input(self) -> Any:
+        return self.dummy_input
```

Comment:
seems there is no test for this evaluator?

Reply:
yes, I think it is bit heavy to test it like other two evaluator, because it needs using a transformer model,
and I am not completely familiar with `Trainer`. my plan is adding a tutorial for pruning Bert with trainer, and add test in that pr.

Reply:
ok